src/lib/base_expression.cpp

```diff
@@ -0,0 +1,398 @@
+  return _left_child;
+}
+
+template <typename DerivedExpressionType>
+void BaseExpression<DerivedExpressionType>::set_left_child(const std::shared_ptr<DerivedExpressionType>& left) {
+  _left_child = left;
+}
+
+template <typename DerivedExpressionType>
+const std::shared_ptr<DerivedExpressionType> BaseExpression<DerivedExpressionType>::right_child() const {
+  return _right_child;
+}
+
+template <typename DerivedExpressionType>
+void BaseExpression<DerivedExpressionType>::set_right_child(const std::shared_ptr<DerivedExpressionType>& right) {
+  _right_child = right;
+}
+
+template <typename DerivedExpressionType>
+ExpressionType BaseExpression<DerivedExpressionType>::type() const {
+  return _type;
+}
+
+template <typename DerivedExpressionType>
+void BaseExpression<DerivedExpressionType>::print(const uint32_t level, std::ostream& out) const {
+  out << std::setw(level) << " ";
+  out << description() << std::endl;
+
+  if (_left_child) {
+    _left_child->print(level + 2u);
+  }
+
+  if (_right_child) {
+    _right_child->print(level + 2u);
+  }
+}
+
+template <typename DerivedExpressionType>
+bool BaseExpression<DerivedExpressionType>::is_operator() const {
+  return is_arithmetic_operator() || is_logical_operator();
+}
+
+template <typename DerivedExpressionType>
+bool BaseExpression<DerivedExpressionType>::is_arithmetic_operator() const {
+  switch (_type) {
+    case ExpressionType::Subtraction:
+    case ExpressionType::Addition:
+    case ExpressionType::Multiplication:
+    case ExpressionType::Division:
+    case ExpressionType::Modulo:
+    case ExpressionType::Power:
+      return true;
+    default:
+      return false;
+  }
+}
+
+template <typename DerivedExpressionType>
+bool BaseExpression<DerivedExpressionType>::is_logical_operator() const {
+  switch (_type) {
+    case ExpressionType::Equals:
+    case ExpressionType::NotEquals:
+    case ExpressionType::LessThan:
+    case ExpressionType::LessThanEquals:
+    case ExpressionType::GreaterThan:
+    case ExpressionType::GreaterThanEquals:
+    case ExpressionType::Like:
+    case ExpressionType::NotLike:
+    case ExpressionType::And:
+    case ExpressionType::Or:
+    case ExpressionType::Between:
+    case ExpressionType::Not:
+    case ExpressionType::Exists:
+      return true;
+    default:
+      return false;
+  }
+}
+
+template <typename DerivedExpressionType>
+bool BaseExpression<DerivedExpressionType>::is_binary_operator() const {
+  if (is_arithmetic_operator()) return true;
+
+  switch (_type) {
+    case ExpressionType::Equals:
+    case ExpressionType::NotEquals:
+    case ExpressionType::LessThan:
+    case ExpressionType::LessThanEquals:
+    case ExpressionType::GreaterThan:
+    case ExpressionType::GreaterThanEquals:
+    case ExpressionType::Like:
+    case ExpressionType::NotLike:
+    case ExpressionType::And:
+    case ExpressionType::Or:
+    case ExpressionType::Between:
```

Comment:
Between takes two arguments, that's why - it has two children. And it's not the same as is_logical_operator because is_binary_operator also cover +,-,*,/

Reply:
(wouldn't argue to much about it, I want Expression gone anyway, soon.)

Reply:
I think I was unclear here. I wanted to say that most of the cases here are covered by `is_logical_operator`. Adding `if (is_logical_operator(...)) return true` would save a lot of lines.

between takes three arguments (a < x < b).